providers/implementations/keymgmt/mlkem_kmgmt.c

```diff
@@ -0,0 +1,585 @@
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    if ((gctx = OPENSSL_zalloc(sizeof(*gctx))) != NULL) {
+        gctx->provctx = provctx;
+        gctx->selection = selection;
+    }
+    if (!mlkem_gen_set_params(gctx, params)) {
+        OPENSSL_free(gctx);
+        gctx = NULL;
+    }
+    debug_print("MLKEMKM gen init returns %p\n", gctx);
+    return gctx;
+}
+
+static int mlkem_gen_set_params(void *genctx, const OSSL_PARAM params[])
+{
+    struct mlkem_gen_ctx *gctx = genctx;
+
+    if (gctx == NULL)
+        return 0;
+
+    debug_print("MLKEMKM empty gen_set params called for %p\n", gctx);
+    return 1;
+}
+
+static const OSSL_PARAM *mlkem_gen_settable_params(ossl_unused void *genctx,
+                                                   ossl_unused void *provctx)
+{
+    static OSSL_PARAM settable[] = {
+        OSSL_PARAM_END
+    };
+    return settable;
+}
+
+static void *mlkem_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
+{
+    struct mlkem_gen_ctx *gctx = (struct mlkem_gen_ctx *)vctx;
+    MLKEM768_KEY *mkey;
+
+    debug_print("MLKEMKM gen called for %p\n", gctx);
+    if (gctx == NULL)
+        return NULL;
+
+    if ((mkey = mlkem_new(gctx->provctx)) == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_INTERNAL_ERROR);
+        return NULL;
+    }
+
+    /* If we're doing parameter generation then we just return a blank key */
+    if ((gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0) {
+        debug_print("MLKEMKM gen returns blank %p\n", mkey);
+        return mkey;
+    }
+
+    mkey->keytype = MLKEM_KEY_TYPE_768;
+
+    if (mkey->encoded_pubkey == NULL) {
+        mkey->encoded_pubkey = OPENSSL_malloc(OSSL_MLKEM768_PUBLIC_KEY_BYTES);
+        if (mkey->encoded_pubkey == NULL) {
+            OPENSSL_free(mkey);
+            return NULL;
+        }
+    }
+
+    ossl_mlkem768_generate_key(mkey->encoded_pubkey, NULL, &mkey->seckey,
+                               mkey->mlkem_ctx);
+    mkey->seckey_initialized = 1;
+    ossl_mlkem768_public_from_private(&mkey->pubkey, &mkey->seckey);
+    mkey->pubkey_initialized = 1;
+
+    debug_print("MLKEMKM gen returns set %p\n", mkey);
+    return mkey;
+}
+
+static void mlkem_gen_cleanup(void *genctx)
+{
+    struct mlkem_gen_ctx *gctx = genctx;
+
+    debug_print("MLKEMKM gen cleanup for %p\n", gctx);
+    OPENSSL_free(gctx);
+}
+
+static void *mlkem_dup(const void *vsrckey, int selection)
+{
+    const MLKEM768_KEY *srckey = (const MLKEM768_KEY *)vsrckey;
+    MLKEM768_KEY *dstkey;
+
+    debug_print("MLKEMKM dup called for %p\n", srckey);
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    dstkey = mlkem_new(srckey->provctx);
+    if (dstkey == NULL)
+        return NULL;
+
+    dstkey->keytype = srckey->keytype;
+    if (srckey->pubkey_initialized == 1
+            && (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
+        /* TODO(ML-KEM)/TBC: By commenting out these, the EVP level test still passes: WHY?? */
```

Comment:
Perhaps the pubkey value from the duplicated key is not really used?

Reply:
The duplication is done during the EVP_PKEY_copy_parameters() call. That call does not set OSSL_KEYMGMT_SELECT_PUBLIC_KEY. So the current test does not exercise this code at all.